inputremapper/injection/consumers/mapping_parser.py

```diff
@@ -0,0 +1,212 @@
+    # all available mapping_handlers
+    "combination": CombinationHandler,
+}
+
+
+def parse_mapping(mapping: Mapping, context: ContextProtocol) -> MappingHandlers:
+    """create dict with a list of MappingHandlers for each key
+
+    Key is of len 1
+    """
+
+    # create a handler for each mapping
+    combination_handlers = {}
+    normal_handlers = {}
+
+    for key, sub_mapping in mapping:
+        config = _create_new_config(key, sub_mapping)
+        if config is None:
+            continue
+
+        _key = Key(config["key"])
+        handler_type = config["type"]
+        if handler_type == "combination":
+            combination_handlers[_key] = _create_handler(config, context)
+        else:
+            assert _key not in normal_handlers.keys()
+            normal_handlers[_key] = _create_handler(config, context)
+
+    # combine all combination handlers such that there is only one handler per single key
+    # if multiple handlers contain the same key, a Hierarchy handler will be created
+    hierarchy_handlers = _create_hierarchy_handlers(combination_handlers)
+    handlers = {}
+    for key, handler in hierarchy_handlers.items():
+        assert len(key) == 1
+        if key[0][0] == EV_KEY:
+            logger.debug("created mapping handler:")
+            logger.debug_mapping_handler(handler)
+            handlers[key] = [handler]
+            continue
+        if key[0][0] == EV_ABS:
+            handler = AbsToBtnHandler(handler, trigger_percent=key[0][2], key=key)
+            logger.debug("created mapping handler:")
+            logger.debug_mapping_handler(handler)
+            handlers[key] = [handler]
+            continue
+        if key[0][0] == EV_REL:
+            handler = RelToBtnHandler(handler, trigger_point=key[0][2], key=key)
+            logger.debug("created mapping handler:")
+            logger.debug_mapping_handler(handler)
+            handlers[key] = [handler]
+            continue
+
+    for key, handler in normal_handlers.items():
+        assert len(key) == 1
+        if key in handlers.keys():
+            logger.debug("created mapping handler:")
+            logger.debug_mapping_handler(handler)
+            handlers[key].append(handler)
+        else:
+            logger.debug("created mapping handler:")
+            logger.debug_mapping_handler(handler)
+            handlers[key] = [handler]
+
+    return handlers
+
+
+def _create_handler(config: Dict[str, any], context) -> MappingHandler:
+    """return the MappingHandler"""
+    return mapping_handler_classes[config["type"]](config, context)
+
+
+def _create_hierarchy_handlers(handlers: Dict[Key, MappingHandler]) -> Dict[Key, MappingHandler]:
+    """sort handlers by sub_keys and create Hierarchy handlers"""
+    # gather all single keys in all mappings
+    sorted_handlers = {}
+    all_keys = handlers.keys()
+    keys = set()  # set of keys with len 1
+    for key in all_keys:
+        for sub_key in key:
```

Comment:
do you have an idea for a better name for the `Key` class? Since it contains one or more keys the name is not very fitting.

I also don't like the term `sub_key`, but I just couldn't come up with something better back then.

How about there is a `class Combination` that contains multiple of `class Key` instead? By doing so, one could write `combination[0].value`, which is much easier to understand than `key[0][0]`

Reply:
Yeah it is currently super confusing I like the idea of using two different classes. The new injection actually depends on the fact that there is a combination, which contains one or more keys, and a key which only has len 1.
I would avoid `Key`, it is way to ambiguous and clashes with python method names like `dict.keys()`

What about `EventCombination` and `EventTuple`

Reply:
I think there would be a bit of an ambiguity between `EventTuple` and `evdev.InputEvent`.

I'd use `evdev.InputEvent` inside `EventCombination`, but it requires a timestamp to be set.

```py
class EventCombination:
    def add_event(type, code, value, target):
        self._events.append(evdev.InputEvent(
            sec=0,
            usec=0,
            type=type,
            code=code,
            value=value
        ))

    def __hash__(self):
        return hash(((event.type, event.code, event.value) for event in self._events))
```

If we never use the evdev.InputEvent constructor directly, this is not an issue. We can just insert dummy numbers for the timestamp.

Reply:
No, they would not work:
```
>> e1 = evdev.InputEvent(0,0,1,5,1)
>> e2 = evdev.InputEvent(0,0,1,5,1)
>> e1 == e2
False
>> e1.__hash__() == e2.__hash__()
False
```
we need both of these properties to be True

Reply:
but

```py
event_combination_1 = EventCombination()
event_combination_1.add_event(1, 5, 1)

event_combination_2 = EventCombination()
event_combination_2.add_event(1, 5, 1)

hash(event_combination_1) == hash(event_combination_2)
```

would

Reply:
No, I currently use a Key of length 1 for all handlers in `context._handlers` and `context.callbacks`.
But this is confusing because it is not clear that we cannot have a combination there. I would rather use some identifier that is distinct from the Identifier which can be a combination.

Reply:
I think a dataclass should do everything we need for us

```
from dataclasses import dataclass
>>> @dataclass
... class EventIdentifier:
...     type: int
...     code: int
...     value: int
    
>>> e1 = EventIdentifier(1,5,1)
>>> e2 = EventIdentifier(1,5,1)
>>> e1 == e2
True
>>> e1.__hash__ == e2.__hash__
True
```

Reply:
> it is not clear that we cannot have a combination there

I see

Or we patch evdev.InputEvent to support __eq__

```py
>>> import evdev
>>> from evdev.ecodes import EV_KEY, KEY_A
>>> 
>>> def __eq__(self, other):
...     return (self.type, self.code, self.value) == (other.type, other.code, other.value)
... 
>>> evdev.InputEvent.__eq__ = __eq__
>>> 
>>> print(hash(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1)))
8742126302377
>>> print(hash(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1)))
8742126302377
>>> print(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1) == evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1))
True
>>> print(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1) == evdev.InputEvent(0, 0, EV_KEY, KEY_A, 0))
False
>>> 
```

Reply:
Ah no we also need to patch `__hash__`, otherwise all InputEvent objects have the same hash:

```py
#!/usr/bin/python

import evdev
from evdev.ecodes import EV_KEY, KEY_A

def __eq__(self, other):
    return (self.type, self.code, self.value) == (other.type, other.code, other.value)

def __hash__(self):
    return hash((self.type, self.code, self.value))

evdev.InputEvent.__eq__ = __eq__
evdev.InputEvent.__hash__ = __hash__

print(hash(evdev.InputEvent(1000, 1000, EV_KEY, KEY_A, 1)))
print(hash(evdev.InputEvent(2000, 3000, EV_KEY, KEY_A, 1)))
print(hash(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 2)))
print(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1) == evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1))
print(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1) == evdev.InputEvent(0, 0, EV_KEY, KEY_A, 0))
```

```
print(hash(evdev.InputEvent(1000, 1000, EV_KEY, KEY_A, 1)))
print(hash(evdev.InputEvent(2000, 3000, EV_KEY, KEY_A, 1)))
print(hash(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 2)))
print(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1) == evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1))
print(evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1) == evdev.InputEvent(0, 0, EV_KEY, KEY_A, 0))

If we use evdev.InputEvent, indexing datastructures with incoming events like

```
something = {
    evdev.InputEvent(0, 0, EV_KEY, KEY_A, 1): "foo"
}

event = device.read_one()
print(something[event])  # "foo"
```

would be possible

Reply:
I don't like the idea of patching evdev or any other module. We don't know how evdev works internally. It feels a bit error prone. Also it is confusing if InputEvent does not behave how it is documented in the evdev documentation.

I would rather create a new class. That way we have full control over the behavior we can give it some useful functions like:
```
@dataclass
class EventData:
    type: int
    code: int
    value: int
    
    def match_input_event(self, event: evdev.InputEvent) -> bool:
       ...

    @classmethod
    def create_from_event(cls, event: evdev.InputEven) -> EventData:
        ...
```

Reply:
I think seems really confusing that there would be an object that holds the exact same information as an `evdev.InputEvent`, has "Event" in the name, but is not actually an event and something different from `evdev.InputEvent`. But I can't come up with good alternatives.

What about converting the `evdev.InputEvent` into our own class (which inherits from `evdev.InputEvent`) after reading? Whenever we see code that goes like `foo.value` we know that this must be an `inputremapper.InputEvent` object. It would have `__hash__` and `__eq__`, we are not patching evdev, and there is no confusion between two very similar classes. Downside is that we have to remember to use that one instead of the one from python-evdev.

Reply:
we are already wrapping evdev.UInput, so the idea is not entirely new

Reply:
Replacing all events with our own version seems like a good idea. We would convert them directly in the `async_read_loop()` and can override different functionality without affecting evdev.

I don't think we need to inherit from `evdev.InputEvent`, since that does not provide any useful functionality.

I would even argue that we should not inherit from `evdev.InputEvent`, instead create it with `@dataclass(frozen=True)` that way we can not change a object after creation and we never need to worry if we should copy a event before we pass it to a function.

Reply:
great! Lets do it